### PR TITLE
feat: add feature for detecting attribute ordering changes

### LINKED ISSUE #225: Attribute order changes neither shown as semantic nor as formatting change
First of all, amazing library. It's really hard to find an XML editing library with these diff capabilities in the Java ecosystem and it works wonderfully.

In my case I need to preserve attribute order, which I found is supported here because attributes are saved in a `LinkedHashMap`.

When pretty printing the attribute order is still preserved, which is what I need in my case.

However I wanted to test some things so I built assertions with the `XmlDiff` utility. I then found that there's a small issue for me. For attribute reordering there are no changes found at all.

It may be that I didn't find the right way to diff things, but since the data model makes this possible, I was wondering if it would be possible to add this somehow?

Ideally with a hasOrderChanges / hasNonSemanticOrderChanges, since I don't care about whitespace changes.

### PR DESCRIPTION
**Pull request by SoftwareApe**
Closes #225 

This would be my proposal for detecting moved attributes.

I added a unit test for it, that makes sure we don't misdetect it as a semantic change now.

Also I wasn't sure if you want to have `ATTRIBUTE_MOVED` as a formatting change or not. Right now !isSemantic -> it's a formatting change.

## Summary by CodeRabbit

* **New Features**
  * XML attribute reordering is now detected and reported as a formatting-only (non-semantic) change.
  * Added a public query to check whether a diff contains attribute order changes.

* **Tests**
  * Added tests verifying detection of attribute reordering and that such changes are non-semantic.

[Review Change Stack](https://app.coderabbit.ai/change-stack/maveniverse/domtrip/pull/226)

### CODE REVIEW

File: core/src/main/java/eu/maveniverse/domtrip/XmlDiff.java [hunk 1]
```diff
@@ -183,6 +183,50 @@ private static void compareAttributes(Element before, Element after, String path
...
+        // changers we first filter both
+        // for the same keys.
+        Set<String> commonKeys = new LinkedHashSet<>(beforeAttrs.keySet());
+        commonKeys.retainAll(afterAttrs.keySet());
+
+        // Build ordered lists (keySet of LinkedHashMap is ordered) of common attributes
+        List<String> beforeCommonKeys =
+                beforeAttrs.keySet().stream().filter(commonKeys::contains).collect(Collectors.toList());
+        List<String> afterCommonKeys =
+                afterAttrs.keySet().stream().filter(commonKeys::contains).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (SoftwareApe) [reply core/src/main/java/eu/maveniverse/domtrip/XmlDiff.java:209]: Value changes should be detected by the preceding code, we explicitly only want to detect moves here.

**Comment by SoftwareApe** [comment]
I have docstrings on all public methods I added, So I'm not sure what the code rabbit review is seeing.

File: core/src/main/java/eu/maveniverse/domtrip/XmlDiff.java [hunk 2]
```diff
@@ -183,6 +183,50 @@ private static void compareAttributes(Element before, Element after, String path
...
+
+    private static void detectAttributeOrderChanges(
+            Map<String, Attribute> beforeAttrs,
+            Map<String, Attribute> afterAttrs,
+            String path,
+            Element before,
+            Element after,
+            List<XmlChange> changes) {
+        // We might have different length of lists and / or different attributes, to not double-detect these as order
+        // changers we first filter both
```
Comment (SoftwareApe) [inline core/src/main/java/eu/maveniverse/domtrip/XmlDiff.java]:
```suggestion
        // changes we first filter both
```

File: core/src/main/java/eu/maveniverse/domtrip/DiffResult.java [hunk 3]
```diff
@@ -158,6 +158,24 @@ public boolean hasFormattingChanges() {
...
+     * <p>Attribute order is not considered a semantic difference in XML, but in
+     * many real-world use cases (configuration files, generated build metadata
+     * etc.) it is desirable to detect reordering separately from other
+     * formatting-only changes like whitespace or quote styles.</p>
+     *
+     * <p>This method reports {@code true} when the diff contains an
+     * {@link ChangeType#ATTRIBUTE_MOVED} change.</p>
+     *
+     * @return {@code true} if there are attribute order changes
+     * @since 1.4.0
```
Comment (SoftwareApe) [inline core/src/main/java/eu/maveniverse/domtrip/DiffResult.java]: Not sure about this versioning tag, it would I guess be on 1.5.0 since it modifies the API.

**Review by gnodet — COMMENTED** [review]
Nice contribution — well-scoped and cleanly integrated. A few items to address before merge:

### 1. `@since` version tag

`DiffResult.java` — `hasAttributeOrderChanges()` is tagged `@since 1.4.0`, but 1.4.0 has already been released. This should be `@since 1.5.0`.

### 2. `ATTRIBUTE_MOVED` + `ATTRIBUTE_CHANGED` overlap

`XmlDiff.java` — `detectAttributeOrderChanges()` builds `commonKeys` based on key presence only, not value equality. If an attribute is both reordered AND its value changed, the code emits both `ATTRIBUTE_CHANGED` and `ATTRIBUTE_MOVED`.

Example: `<root a="1" b="2"/>` → `<root b="3" a="1"/>` produces `ATTRIBUTE_CHANGED` for `b` AND `ATTRIBUTE_MOVED` for both `a` and `b`.

This may be intentionally desired (they are independent facts), but it should be a conscious decision. Please clarify the intended behavior — should `ATTRIBUTE_MOVED` only apply to attributes with unchanged values, or is emitting both correct?

### 3. Test coverage

The single test covers a basic two-attribute swap. A few more cases would strengthen confidence:
- Mixed reorder + value change (the edge case above)
- Mixed reorder + attribute add/remove
- `hasAttributeOrderChanges()` directly asserted (as CodeRabbit also suggested)
- Negative case: same order → no `ATTRIBUTE_MOVED`

---

_Claude Code on behalf of Guillaume Nodet_

**Comment by SoftwareApe** [comment]
@gnodet thank you for your review, I'll adjust these.

* I added the 1.5.0 version tag
* I added some more tests (positive, negative, namespacing) and the one you mentioned about moved and value changed.
* I kept the semantics the same, I do think a change of position and a change of value are two distinct things, but I'm open to discussion here.

File: core/src/main/java/eu/maveniverse/domtrip/DiffResult.java (hunk 3 above)
Comment (SoftwareApe) [reply core/src/main/java/eu/maveniverse/domtrip/DiffResult.java]: Changed to 1.5.0 as suggested.

File: core/src/test/java/eu/maveniverse/domtrip/XmlDiffTest.java [hunk 4]
```diff
@@ -554,6 +558,80 @@ void changesUnderUsesBoundaryAwareMatching() {
...
+        assertTrue(
+                result.changes().stream().anyMatch(c -> c.type() == ChangeType.ATTRIBUTE_MOVED),
+                "Expected ATTRIBUTE_MOVED for reordered attributes");
+        assertTrue(result.hasAttributeOrderChanges(), "Expected hasAttributeOrderChanges() to be true");
+
+        if (expectSemanticChange) {
+            assertTrue(result.hasSemanticChanges(), "Expected semantic changes for this scenario");
+        } else {
+            assertFalse(result.hasSemanticChanges(), "Attribute reordering alone should not be semantic");
+        }
```
[the remark this answers is not in the document]
Comment (SoftwareApe) [reply core/src/test/java/eu/maveniverse/domtrip/XmlDiffTest.java:595]: Doesn't work well with the parametrized test, also retests attribute changed functionality that is not what we're testing for here, rather we want to see that it detects the semantic change as usual while also detecting the attribute moved change.
